bin/reth/src/cli/node_builder.rs

```diff
@@ -0,0 +1,1336 @@
+use once_cell::sync::Lazy;
+use reth_auto_seal_consensus::{AutoSealBuilder, AutoSealConsensus, MiningMode};
+use reth_beacon_consensus::{
+    hooks::{EngineHooks, PruneHook},
+    BeaconConsensus, BeaconConsensusEngine, BeaconConsensusEngineError,
+    MIN_BLOCKS_FOR_PIPELINE_RUN,
+};
+use reth_blockchain_tree::{
+    config::BlockchainTreeConfig, externals::TreeExternals, BlockchainTree, ShareableBlockchainTree,
+};
+use reth_config::{
+    config::{PruneConfig, StageConfig},
+    Config,
+};
+use reth_db::{
+    database::Database,
+    database_metrics::{DatabaseMetadata, DatabaseMetrics},
+};
+use reth_downloaders::{
+    bodies::bodies::BodiesDownloaderBuilder,
+    headers::reverse_headers::ReverseHeadersDownloaderBuilder,
+};
+use reth_interfaces::{
+    blockchain_tree::BlockchainTreeEngine,
+    consensus::Consensus,
+    p2p::{
+        bodies::{client::BodiesClient, downloader::BodyDownloader},
+        either::EitherDownloader,
+        headers::{client::HeadersClient, downloader::HeaderDownloader},
+    },
+    RethResult,
+};
+use reth_network::{NetworkBuilder, NetworkConfig, NetworkEvents, NetworkHandle, NetworkManager};
+use reth_network_api::{NetworkInfo, PeersInfo};
+use reth_primitives::{
+    constants::eip4844::{LoadKzgSettingsError, MAINNET_KZG_TRUSTED_SETUP},
+    kzg::KzgSettings,
+    stage::StageId,
+    BlockHashOrNumber, BlockNumber, ChainSpec, DisplayHardforks, Head, SealedHeader, TxHash, B256,
+    MAINNET,
+};
+use reth_provider::{
+    providers::BlockchainProvider, BlockHashReader, BlockReader,
+    BlockchainTreePendingStateProvider, CanonStateSubscriptions, HeaderProvider, HeaderSyncMode,
+    ProviderFactory, StageCheckpointReader,
+};
+use reth_prune::PrunerBuilder;
+use reth_revm::EvmProcessorFactory;
+use reth_revm_inspectors::stack::Hook;
+use reth_rpc_engine_api::EngineApi;
+use reth_stages::{
+    prelude::*,
+    stages::{
+        AccountHashingStage, ExecutionStage, ExecutionStageThresholds, IndexAccountHistoryStage,
+        IndexStorageHistoryStage, MerkleStage, SenderRecoveryStage, StorageHashingStage,
+        TotalDifficultyStage, TransactionLookupStage,
+    },
+    MetricEvent,
+};
+use reth_tasks::{TaskExecutor, TaskManager};
+use reth_transaction_pool::{
+    blobstore::InMemoryBlobStore, EthTransactionPool, TransactionPool,
+    TransactionValidationTaskExecutor,
+};
+use secp256k1::SecretKey;
+use std::{
+    net::{SocketAddr, SocketAddrV4},
+    path::PathBuf,
+    sync::Arc,
+};
+use tokio::{
+    runtime::Handle,
+    sync::{
+        mpsc::{unbounded_channel, Receiver, UnboundedSender},
+        oneshot, watch,
+    },
+};
+use tracing::*;
+
+/// The default prometheus recorder handle. We use a global static to ensure that it is only
+/// installed once.
+pub static PROMETHEUS_RECORDER_HANDLE: Lazy<PrometheusHandle> =
+    Lazy::new(|| prometheus_exporter::install_recorder().unwrap());
+
+/// Start the node
+#[derive(Debug)]
+pub struct NodeBuilder {
```

Comment:
I think we eventually want to rename this type to NodeConfig and have a generic NodeBuilder that takes the config

Reply:
renamed, kept `NodeBuilderWithDatabase`